design-proposals/klip-19-materialize-views.md

```diff
@@ -0,0 +1,223 @@
+including another KSQL query, should be treated as _read-only_ as KSQL can not offer correct table semantics on `INSERT`s.
+
+SQL already has concepts to differentiate: 
+ * a `TABLE` is a source dataset. The data in the table can be mutated via `INSERT`/`UPDATE`/`DELETE`s.
+ * a `VIEW` is a derived and read-only dataset. It's state is _derived_ from one or more sources. 
+  `INSERT`/`UPDATE`/`DELETE`s are not supported. 
+ * A `MATERIALIZED VIEW` is a `VIEW` where the cost of calculating the derived dataset is done on the write path, rather 
+   than the read.
+
+We propose introducing a `CREATE MATERIALIZED VIEW` statement to allow users to import table datasets which KSQL should
+treat as read-only.  
+
+The existing `CREATE TABLE` statement will be repurposed for defining new tables or importing an existing topic as a table. 
+Note: if `CREATE TABLE` is used to import a topic which is updated by another system then the behaviour is _undefined_, 
+i.e. table semantics can not be guaranteed.
+   
+KSQL currently _materializes_ the results of `CREATE TABLE AS SELECT` statements into the sink topic. 
+Hence we propose renaming `CTAS` to `CREATE MATERIALIZED VIEW AS SELECT` to make it clear this is a _derived_ dataset,
+and can not be updated directly.
+
+The existing `CREATE TABLE AS SELECT` statement will be removed from KSQL syntax and reserved for future use.
+If/when reintroduced it may create a table snapshot, i.e. a `TABLE`, not a `VIEW`, that is initialized from the 
+supplied query, i.e. a one-time initialization, not a continuous query/update. Though such functionality may 
+be better expressed using SQL `INSERT INTO` syntax. Alternatively, `CTAS` could be reintroduced as a continuously 
+updating table if we were to also offer correct `INSERT`, and later `UPDATE` and `DELETE` semantics, which may be possible.
+
+In case you spotted the edge-case, let's discuss it: what about the case where the data in Kafka is the _source dataset_, 
+but its managed by another system? As we know, KSQL can not offer correct `INSERT` table semantics, but it's still a 
+_source_, not a _derived dataset_, right? So how does this fit in the `TABLE` vs `VIEW` world? Well, KSQL may be reading 
+from the _source dataset_, but its essentially making a copy when it reads the data. That copy is a `VIEW`. Hence, users
+should import such a data set using `CREATE MATERIALIZED VIEW`.
+
+### What about streams? 
+
+Streams don't have the complexities of per-key update semantics that tables have. So, it's OK to insert into them, right?
+
+Well, streams are certainly not as clear cut as tables. But maybe there's a case for differentiating between a stream and
+a read-only stream view. 
+
+Consider a stream coming from some upstream system, e.g a click stream from a web-tier. Should a KSQL user be able to 
+add rows to such a stream? Probably not, but the true answer is most likely 'it depends'. It depends on how the organisation
```

Comment:
With the new work on KLIP-15 - creating a new API and client, we are explicitly supporting the use case of inserting rows into streams. We want apps (including web apps) to use the new ksqlDB clients (or use the new HTTP2 API directly) to write apps including "web tiers". This is a key point for a great user experience as we don't want apps to also have to manage a Kafka client to insert events into their click stream.
Some users will continue to use Kafka directly for their click stream but I don't think we should discourage users from the new way which we're strategically targeting. (I.e. "Probably not" seems too strong) ;)

Reply:
Yep agree. It does go on to explain 'it depends'. But I've reworded slightly.